tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -848,3 +848,48 @@ def prepare_r_uls_r_lus(
             r_lu_matrix,
         )
         # TODO: beta sobolev needs to be recalculated for each iteration, because it depends on number density
+
+    @staticmethod
+    def create_coll_exc_deexc_matrix(
+        coll_exc_coefficient, coll_deexc_coefficient
+    ):
+        """Generates a coefficient matrix from collisional excitation/deexcitation coefficients.
+
+        Needs to be multiplied by electron density when added to the overall rate_matrix.
+        Parameters
+        ----------
+        coll_exc_coefficient : pandas.DataFrame
+            DataFrame of collisional excitation coefficients for current (atomic number, ion_number)
+        coll_deexc_coefficient : pandas.DataFrame
+            DataFrame of collisional deexcitation coefficients for (atomic number, ion_number)
+
+        Returns
+        -------
+        coeff_matrix : np.array (number of levels, number of levels)
+            Square matrix constructed by collisional exc./deexc. coefficients.
+        """
+        size = (
+            coll_exc_coefficient.index.get_level_values("level_number_lower")
+            .unique()
+            .size
+        )
+        diagonal_exc = coll_exc_coefficient.groupby("level_number_lower").sum()
+        diagonal_deexc = coll_deexc_coefficient.groupby(
+            "level_number_upper"
+        ).sum()
+        exc_matrix = np.zeros((size + 1, size + 1))
+        deexc_matrix = np.zeros((size + 1, size + 1))
+        exc_matrix[
+            np.tril_indices(size + 1, k=-1)
+        ] = coll_exc_coefficient.values.reshape(
+            size + 1,
+        )
+        deexc_matrix[
+            np.triu_indices(size + 1, k=1)
+        ] = coll_deexc_coefficient.values.reshape(
+            size + 1,
+        )
```

Comment:
This assumes that the index for  `coll_exc_coefficient` and `coll_deexc_coefficient` are sorted (which I guess they should be) and that there are no missing values. It might be a bit safer and simpler to just use direct indexing e.g.: `exc_matrix[(coll_exc_coefficient.index.get_level_values('level_number_upper'), coll_exc_coefficient.index.get_level_values('level_number_lower'))] = coll_exc_coefficient.values`